calico_upgrade/pkg/migrate/felixconfig.go

```diff
@@ -0,0 +1,279 @@
+	// transfer the same value across.
+	clusterInfo := apiv3.NewClusterInformation()
+	clusterInfo.Name = "default"
+	if err = fc.parseFelixConfigV1IntoResourceV3(kvps, clusterInfo, data); err != nil {
+		return err
+	}
+
+	// Query all of the per-host felix config into a slice of KVPairs.
+	kvps, err = clientv1.List(model.HostConfigListOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Sort the configuration into slices of KVPairs for each node, converting the
+	// nodename as we go.
+	nodeKvps := make(map[string][]*model.KVPair, 0)
+	for _, kvp := range kvps {
+		// Extract the key, update it and store the updated key. Store in the node-specific
+		// bucket.
+		hk := kvp.Key.(model.HostConfigKey)
+		hk.Hostname = conversionv1v3.ConvertNodeName(hk.Hostname)
+		kvp.Key = hk
+
+		nodeKvps[hk.Hostname] = append(nodeKvps[hk.Hostname], kvp)
+	}
+
+	// For each node, get the felix config kvps and convert to v3 per-node
+	// FelixConfiguration resource.
+	for node, kvps := range nodeKvps {
+		// Convert to v3 resource.
+		nodeConfig := apiv3.NewFelixConfiguration()
+		nodeConfig.Name = fmt.Sprintf("node.%s", node)
+		if err := fc.parseFelixConfigV1IntoResourceV3(kvps, nodeConfig, data); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// This function converts a slice of v1 KVPairs into the appropriate v3 values and
+// merges the results into a single v3 resource Spec for felix configuration (global
+// or per host) or a clusterInfo.
+// Conversion errors are added to the ConvertedData struct.
+func (fc *felixConfig) parseFelixConfigV1IntoResourceV3(
+	kvps []*model.KVPair,
+	res conversionv1v3.Resource,
+	data *ConvertedData,
+) error {
+	logCxtRes := log.WithFields(log.Fields{
+		"kind": res.GetObjectKind().GroupVersionKind().Kind,
+		"name": res.GetObjectMeta().GetName(),
+	})
+
+	// Convert the KVP slice into a name value map.
+	config := map[string]string{}
+	for _, kvp := range kvps {
+		if kvp.Value == nil {
+			continue
+		}
+		switch key := kvp.Key.(type) {
+		case model.GlobalConfigKey:
+			config[key.Name] = kvp.Value.(string)
+		case model.HostConfigKey:
+			config[key.Name] = kvp.Value.(string)
+		}
+	}
+
+	// Extract the Spec from the resource FelixConfiguration or ClusterInfo.
+	specValue := reflect.ValueOf(res).Elem().FieldByName("Spec")
+	if !specValue.IsValid() {
+		return fmt.Errorf("unable to process config resource type: %v", res)
+	}
+
+	// Loop through the Spec setting each field from the supplied KVPair data.
+	setField := false
+	specType := specValue.Type()
+	for i := 0; i < specType.NumField(); i++ {
+		field := specType.Field(i)
+		fieldValue := specValue.Field(i)
+
+		// Get the v1 config value associated with the field.
+		configName := fc.getConfigName(field)
+		logCxt := logCxtRes.WithFields(log.Fields{
+			"field":  field.Name,
+			"config": configName,
+		})
+		configStrValue, ok := config[configName]
+		if !ok {
+			logCxt.Debug("config value is not configured in v1")
```

Comment:
helpful to report the `configName` here? Also is this common/important enough to warrant inclusion in a report?

Reply:
This is included in the logCxt.